_posts/2017-10-26-totw-94.md

```diff
@@ -0,0 +1,131 @@
+---
+title: "Tip of the Week #94: Callsite Readability and bool Parameters"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/94
+type: markdown
+order: 094
+---
+
+Originally posted as totw/94 on 2015-04-27
+
+*By Geoff Romer [(gromer@google.com)](mailto:gromer@google.com)*
+
+Revised 2017-10-25
+
+*"Of the many forms of false culture, a premature converse with abstractions is
+perhaps the most likely to prove fatal to the growth of &hellip; vigour of
+intellect." &mdash; George Boole.*
+
+Suppose you come across code like this:
+
+```c++
+int main(int argc, char* argv[]) {
+  ParseCommandLineFlags(&argc, &argv, false);
+}
+```
+
+Can you tell what this code does, and in particular what the last argument
+means? Now suppose you've come across this function before, and you know that
+the last argument has to do with whether command-line flags are left in argv
+after the call. Can you tell which is true and which is false?
+
+Of course you don't know, because this is hypothetical, but even in real code,
+we have better things to do with our brains than memorize the meaning of every
+function parameter, and better things to do with our time than go look up the
+documentation for every function call we encounter. We have to be able to make
+pretty good guesses about what function calls mean, just from looking at the
+callsite.
+
+Well-chosen function names are the key to making function calls readable, but
+they're often not enough. We usually need the arguments themselves to give us
+some clues about what they mean. For example, you might not know what to make of
+`absl::string_view s(x, y);` if you'd never seen `string_view` before, but
+`absl::string_view s(my_str.data(), my_str.size());` and `absl::string_view
+s("foo");` are much clearer. The trouble with `bool` parameters is that the
+argument at the callsite is very often a literal `true` or `false`, and that
+gives the reader no contextual cues about the meaning of the parameter, as we
+saw in the `ParseCommandLineFlags` example. This problem is compounded if
+there's more than one `bool` parameter, because now you have the additional
+problem of figuring out which parameter is which.
+
+So how can we fix code like that example? One possibility is to do something
+like this:
+
+```c++ {.bad}
```

Comment:
Remove {.bad} It's not supported in our GH markdown

Reply:
:-(

We should figure out some sort of convention for distinguishing good from bad code, then. I don't think we can just drop the tags without losing important information.

Reply:
Add a comment?